Mathlib/LinearAlgebra/InvariantBasisNumber.lean

```diff
@@ -11,34 +12,65 @@ import Mathlib.RingTheory.PrincipalIdealDomain
+  of a finitely generated free (left) `R`-module.
 
-## Main definitions
+It is also useful to consider the following stronger conditions:
+
+- the *rank condition*, witnessed by the type class `RankCondition R`, states that
+  the existence of a surjective linear map `(Fin n → R) →ₗ[R] (Fin m → R)` implies `m ≤ n`
+
+- the *strong rank condition*, witnessed by the type class `StrongRankCondition R`, states
+  that the existence of an injective linear map `(Fin n → R) →ₗ[R] (Fin m → R)`
+  implies `n ≤ m`.
+
+- `OrzechProperty R`, defined in `Mathlib/RingTheory/OrzechProperty.lean`,
+  is a type class stating that `R` satisfies the following property:
+  for any finitely generated `R`-module `M`, any surjective homomorphism `f : N → M`
+  from a submodule `N` of `M` to `M` is injective.
+  It was introduced in papers by Orzech [orzech1971], Djoković [djokovic1973] and
+  Ribenboim [ribenboim1971], under the names `Π`-ring or `Π₁`-ring.
+  It implies the strong rank condition if the ring is nontrivial.
+
+
+## Instances
+
+- `IsNoetherianRing.orzechProperty` (defined in `Mathlib/RingTheory/Noetherian.lean`) :
+  (TODO: in a future PR)
+  any left-noetherian ring satisfies the Orzech property.
+  This applies in particular to division rings.
 
-`StrongRankCondition R` is a type class stating that `R` satisfies the strong rank condition.
-`RankCondition R` is a type class stating that `R` satisfies the rank condition.
-`InvariantBasisNumber R` is a type class stating that `R` has the invariant basis number property.
+- `strongRankCondition_of_orzechProperty` : the Orzech property implies the strong rank condition
+  (for non trivial rings).
 
-## Main results
+- `rankCondition_of_strongRankCondition` : the strong rank condition implies the rank condition.
 
-We show that every nontrivial left-noetherian ring satisfies the strong rank condition,
-(and so in particular every division ring or field),
-and then use this to show every nontrivial commutative ring has the invariant basis number property.
+- `invariantBasisNumber_of_rankCondition` : the rank condition implies the
+  invariant basis number property.
+
+- `invariantBasisNumber_of_nontrivial_of_commRing`: a nontrivial commutative ring satisfies
+  the invariant basis number property
+
+More generally, every commutative ring satisfies the Orzech property and the strong rank condition.
+The corresponding instances are proved in
+`Mathlib/RingTheory/FiniteType.lean` (TODO: in a future PR).
+We keep `invariantBasisNumber_of_nontrivial_of_commRing` here since it imports fewer files.
+
+
+## Counterexamples to converse results
+
+The following examples can be found in the book of Lam [lam_1999]
+(see also <https://math.stackexchange.com/questions/4711904>):
+
+- The free algebra `k⟨x, y⟩` satisfies the rank condition but not the strong rank condition.
+- The ring `ℚ⟨a, b, c, d⟩ / (ac − 1, bd − 1, ab, cd)` satisfies the invariant basis number property
```

Comment:
Just to show my ignorance -- is this a quotient by a 2-sided ideal? And k<x,y> is the free non-abelian algebra?

Reply:
> is this a quotient by a 2-sided ideal?

I think yes. Otherwise the quotient is not a ring.

> k<x,y> is the free non-abelian algebra?

Yes.